### PR TITLE
Disable Schema Compat plugin during server upgrade

### PR DESCRIPTION
**Pull request by rcritten**
If this is enabled it can cause a deadlock with SSSD trying
to look up entries and it trying to get data on AD users
from SSSD.

When reading the entry from LDIF try to get the camel-case
nsslapd-pluginEnabled and fall back to the all lower-case
nsslapd-pluginenabled if that is not found. It would be nice
if the fetch function was case sensitive but this is likely
overkill as it is, but better safe than blowing up.

Upon restoring it will always write the camel-case version.

https://pagure.io/freeipa/issue/6721

Signed-off-by: Rob Crittenden <rcritten@redhat.com>

### CODE REVIEW

File: ipaserver/install/upgradeinstance.py [hunk 1]
```diff
@@ -161,6 +163,29 @@ def __save_config(self):
             else:
                 self.backup_state('nsslapd-global-backend-lock', global_lock)
 
+        with open(self.filename, "r") as in_file:
+            parser = GetEntryFromLDIF(in_file, entries_dn=[COMPAT_DN])
+            parser.parse()
+
+            try:
+                compat_entry = parser.get_results()[COMPAT_DN]
```
Comment (stlaz) [inline ipaserver/install/upgradeinstance.py]: I don't think you still need to have the file opened here.

File: ipaserver/install/upgradeinstance.py [hunk 2]
```diff
@@ -161,6 +163,29 @@ def __save_config(self):
...
+            except KeyError:
+                pass
```
Comment (stlaz) [inline ipaserver/install/upgradeinstance.py]: Just add a `return` here and remove the else below.

File: ipaserver/install/upgradeinstance.py [hunk 3]
```diff
@@ -161,6 +163,29 @@ def __save_config(self):
...
+            else:
+                try:
+                    schema_compat_enabled = compat_entry[
+                        'nsslapd-pluginEnabled'][0].decode('utf-8')
```
Comment (stlaz) [inline ipaserver/install/upgradeinstance.py]: Is the returned entry just a normal dictionary, does python-ldap not set case for all the keys, or doesn't python-ldap have their own caseless dictionary for these cases?

File: ipaserver/install/upgradeinstance.py [hunk 4]
```diff
@@ -161,6 +163,29 @@ def __save_config(self):
...
+                    self.backup_state('schema_compat_enabled',
+                                      schema_compat_enabled)
+                except KeyError:
+                    try:
+                        schema_compat_enabled = compat_entry[
```
Comment (stlaz) [inline ipaserver/install/upgradeinstance.py]: You really should duplicate code less:
```python
schema_compat_enabled = compat_entry.get('nsslapd-pluginEnabled')
if schema_compat_enabled is None:
     schema_compat_enabled = compat_entry.get('nsslapd-pluginenabled')
if schema_compat_enabled:
    self.backup_state('schema_compat_enabled', schema_compat_enabled)
```

File: ipaserver/install/upgradeinstance.py (hunk 2 above)
Comment (rcritten) [reply ipaserver/install/upgradeinstance.py]: Done.

**Comment by rcritten** [comment]
This uses the same process to change the dse.ldif values that has been used for quite some time. I saw no AVCs in F28.

**Comment by stlaz** [comment]
Went again through the code, all LGTM, I trust you there're no AVCs and that it works :stuck_out_tongue:

**Comment by rcritten** [comment]
master:

* 2b3eb5c567e1e5d4fe7d945e36615cef3dd4d144 Disable Schema Compat plugin during server upgrade

**Pull request closed without merging**